### PR TITLE
feat: portals assets generation

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR adds portals assets generation, alongside updated `generatedAssetData.json`.

Note: This does *not* yet add support for Ethereum nor BSC assets, see striked-thru risks below.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- tackles https://github.com/shapeshift/web/issues/7303

Actually low now 
~~Twofolds high:~~

- ~~`generatedAssetData.json` will now be a 30ish MB large boi, and could eventually grow up to 100MB+, in which case we would need git-lfs. Immediate concerns as a result of this are:~~
  - ~~the Cloudflare build, which may be rugged~~
  - ~~much slower generation time, which makes local regeneration pretty much impossible~~
  - ~~Zerion related assets generation won't be so cheap~~
- ~~from a user standpoint, having so many assets in the app makes asset selection notably slower~~

~~If these concerns end up being blockers, a solution may be to land this PR as a `tackles` vs. `closes`, and don't include Ethereum nor BSC assets, all of which have 10k+ assets, until we figure out a solution to this.~~

Turns out they were blockers - disabled Ethereum and BSC Portals assets fetching for the time being. 

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

- Portal assets are present in the app at runtime - try with `Trader Joe` assets
- Current related assets are still happy
- Do a quick smoke tests of major assets and ensure they are still happy

### Engineering

- ☝🏽 
- Diff looks sane, with added Portal assets for
  - Arbitrum
  - Avalanche
  - Base
  - Gnosis
  - Optimism
  - Polygon
  
## Screenshots (if applicable)

image
image

### CODE REVIEW

File: .env.develop [hunk 1]
```diff
@@ -2,7 +2,9 @@
 REACT_APP_FEATURE_RFOX_REWARDS_TX_HISTORY=true
 REACT_APP_FEATURE_CHATWOOT=true
 REACT_APP_FEATURE_ARBITRUM_BRIDGE=true
+REACT_APP_FEATURE_COWSWAP_ARBITRUM=true
```
Comment (gomesalexandre) [inline .env.develop:5]: revert me

File: .env.dev [hunk 2]
```diff
@@ -1,8 +1,10 @@
 # feature flags
 REACT_APP_FEATURE_RFOX_REWARDS_TX_HISTORY=true
 REACT_APP_FEATURE_ARBITRUM_BRIDGE=true
+REACT_APP_FEATURE_COWSWAP_ARBITRUM=true
```
Comment (gomesalexandre) [inline .env.dev:4]: and me too

File: .env.base [hunk 3]
```diff
@@ -40,6 +40,7 @@ REACT_APP_FEATURE_RFOX_REWARDS_TX_HISTORY=false
 REACT_APP_FEATURE_ARBITRUM_BRIDGE=false
 REACT_APP_FEATURE_CUSTOM_TOKEN_IMPORT=false
 REACT_APP_FEATURE_ARBITRUM_BRIDGE_CLAIMS=false
+REACT_APP_FEATURE_PORTALS_SWAPPER=false
```
Comment (gomesalexandre) [inline .env.base:43]: and this guy as well

File: scripts/generateAssetData/ethereum/index.ts [hunk 4]
```diff
@@ -10,10 +10,9 @@ import * as coingecko from '../coingecko'
 import type { IdenticonOptions } from '../generateAssetIcon/generateAssetIcon'
 import { getRenderedIdenticonBase64 } from '../generateAssetIcon/generateAssetIcon'
 import { generateTrustWalletUrl } from '../generateTrustWalletUrl/generateTrustWalletUrl'
+import { getPortalTokens } from '../utils/portals'
 import { getIdleTokens } from './idleVaults'
 import { getUniswapV2Pools } from './uniswapV2Pools'
-// Yearn SDK is currently rugged upstream
-// import { getUnderlyingVaultTokens, getYearnVaults, getZapperTokens } from './yearnVaults'
```
Comment (gomesalexandre) [inline scripts/generateAssetData/ethereum/index.ts:16]: revert comment

File: scripts/generateAssetData/ethereum/index.ts [hunk 5]
```diff
@@ -33,26 +32,17 @@ const foxyToken: Asset = {
 }
 
 export const getAssets = async (): Promise<Asset[]> => {
-  const [ethTokens, uniV2PoolTokens, idleTokens] = await Promise.all([
+  const [ethTokens, uniV2PoolTokens, idleTokens, portalsAssets] = await Promise.all([
     coingecko.getAssets(ethChainId),
-    // getYearnVaults(),
-    // getZapperTokens(),
-    // getUnderlyingVaultTokens(),
```
Comment (gomesalexandre) [inline scripts/generateAssetData/ethereum/index.ts:40]: and here

File: scripts/generateAssetData/ethereum/index.ts [hunk 6]
```diff
@@ -33,26 +32,17 @@ const foxyToken: Asset = {
...
+    getPortalTokens(ethereum),
   ])
 
-  const ethAssets = [
-    ...idleTokens,
-    foxyToken,
-    ...ethTokens,
-    // ...yearnVaults,
-    // ...zapperTokens,
-    // ...underlyingTokens,
```
Comment (gomesalexandre) [inline scripts/generateAssetData/ethereum/index.ts:51]: also here

File: scripts/generateAssetData/utils/portals.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  [gnosisChainId]: 'gnosis',
+  [baseChainId]: 'base',
+}
+
+export const fetchPortalsTokens = async (
+  chainId: ChainId,
+  page: number = 0,
+  accTokens: TokenInfo[] = [],
+): Promise<TokenInfo[]> => {
+  console.log(`Fetching page ${page} of Portals tokens for chainId: ${chainId}`)
```
Comment (gomesalexandre) [inline scripts/generateAssetData/utils/portals.ts:56]: Revert me before opening PR, useful for debugging purposes

Comment (gomesalexandre) [reply scripts/generateAssetData/utils/portals.ts:56]: Actually don't, this is pretty useful to get a grasp of something happening when devving locally, especially with the high amount of requests

File: scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts [hunk 8]
```diff
@@ -288,7 +241,9 @@ export const generateRelatedAssetIndex = async () => {
     drainPerInterval: 25, // Adjusted drain rate to replenish at a sustainable pace
     intervalMs: 2000,
   })
-  for (const batch of chunkArray(Object.keys(generatedAssetData), BATCH_SIZE)) {
+  const chunks = chunkArray(Object.keys(generatedAssetData), BATCH_SIZE)
+  for (const [i, batch] of chunks.entries()) {
+    console.log(`Fetching chunk: ${i} of ${chunks.length}`)
```
Comment (gomesalexandre) [inline scripts/generateAssetData/generateRelatedAssetIndex/generateRelatedAssetIndex.ts:246]: Added for the same reason of getting a grasp of what's happening during this very long regen

File: scripts/generateAssetData/arbitrum/index.ts [hunk 9]
```diff
@@ -1,13 +1,19 @@
...
 import * as coingecko from '../coingecko'
 import { getRenderedIdenticonBase64 } from '../generateAssetIcon/generateAssetIcon'
+import { getPortalTokens } from '../utils/portals'
 
 export const getAssets = async (): Promise<Asset[]> => {
-  const assets = await coingecko.getAssets(arbitrumChainId)
-  return [...assets, arbitrum].map(asset => ({
+  const [assets, portalsAssets] = await Promise.all([
+    coingecko.getAssets(arbitrumChainId),
+    getPortalTokens(arbitrum),
```
Comment (0xApotheosis) [inline scripts/generateAssetData/arbitrum/index.ts:13]: I'm assuming we are happy stopping here if one of those calls fail? I.e. we are happy with `Promise.all` vs `Promise.allSettled`.

File: scripts/generateAssetData/bnbsmartchain/index.ts [hunk 10]
```diff
@@ -1,12 +1,22 @@
...
 import { getRenderedIdenticonBase64 } from '../generateAssetIcon/generateAssetIcon'
+// import { getPortalTokens } from '../utils/portals'
 
-export const getAssets = async () => {
-  const assets = await coingecko.getAssets(bscChainId)
-  return [...assets, bnbsmartchain].map(asset => ({
+export const getAssets = async (): Promise<Asset[]> => {
+  const [assets, portalsAssets] = await Promise.all([
+    coingecko.getAssets(bscChainId),
+    // TODO(gomes): revert me back, there are 10k+ assets for Ethereum = problems
```
Comment (0xApotheosis) [inline scripts/generateAssetData/bnbsmartchain/index.ts:13]:
```suggestion
    // TODO(gomes): revert me back, there are 10k+ assets for BNB = problems
```

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
This is all looking good, except that:

❌ Portals assets are not present in the global search (ignore if this is intentional):

Screenshot 2024-07-23 at 2 17 28 PM

✅ Trade search:

Screenshot 2024-07-23 at 2 16 59 PM

✅ Related assets

Screenshot 2024-07-23 at 2 18 44 PM

File: scripts/generateAssetData/utils/portals.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+  const url = 'https://api.portals.fi/v2/tokens'
+  const PORTALS_API_KEY = process.env.REACT_APP_PORTALS_API_KEY
+
+  const { throttle, clear } = createThrottle({
+    capacity: 100, // 100 requests per minute https://build.portals.fi/billing
```
Comment (suhailgme) [inline scripts/generateAssetData/utils/portals.ts:61]: Fyi: Updated rpm limit to 500 and request limit to 15,000,000 (up from 100 and 50,000). See dashboard for confirmation.

**Comment by gomesalexandre** [comment]
> This is all looking good, except that:
> 
> ❌ Portals assets are not present in the global search (ignore if this is intentional):
> 
> Screenshot 2024-07-23 at 2 17 28 PM

Not intended per se, but I've noticed this as well @0xApotheosis! This is an issue that's currently present in develop, see one of the latest assets in develop's `generatedAssetData.json`. Will capture an issue for it.

image

File: scripts/generateAssetData/utils/portals.ts (hunk 11 above)
Comment (gomesalexandre) [reply scripts/generateAssetData/utils/portals.ts:61]: well why thank you @suhailgme! 🤯 
image

**Comment by 0xApotheosis** [comment]
Ah ok, these assets don't have market data, hey? 
I've noticed that unless an asset's key exists in the market data slice then it won't show in the global asset search.

**Review by 0xApotheosis — APPROVED** [review]
Approving as the global asset search issue is unrelated to this PR 🙏

File: scripts/generateAssetData/arbitrum/index.ts (hunk 9 above)
Comment (gomesalexandre) [reply scripts/generateAssetData/arbitrum/index.ts]: Absolutely in on that one - the reason why it "worked" for Ethereum (which uses `Promise.all()` is Idle throws and doesn't recatch, and the two other Uni-V2/Coingecko asset services "normally work", which as we can probably all agree is definitely not safe 😄 

d59b7f72f9aa3b11b02856547d4bfaef76d0f9

File: scripts/generateAssetData/bnbsmartchain/index.ts (hunk 10 above)
Comment (gomesalexandre) [reply scripts/generateAssetData/bnbsmartchain/index.ts]: 2ccc55cb001695a54dd354737f3f45534245277f

File: scripts/generateAssetData/utils/portals.ts (hunk 11 above)
Comment (gomesalexandre) [reply scripts/generateAssetData/utils/portals.ts]: 3fea3ad6c2143d4d457fa3b76c9a1bc63854c8b8

**Review by gomesalexandre — COMMENTED** [review]
Sanity retest

### Assets page still looks sane

image
 
### Related assets are still happy

image


### Portals assets are still present

image

**Comment by gomesalexandre** [comment]
Disabling auto-merge as this may or may not rug TS LSP - possibly a my machine kind of thing but...

File: tsconfig.json [hunk 12]
```diff
@@ -35,7 +35,7 @@
   "include": ["src/**/*", "src/**/*.json", "packages/**/*", "packages/**/*.json", "scripts/generateAssetData",
     "scripts/generateTradableThorAssetMap"
   ],
-  "exclude": ["dist", "node_modules", "**/__mocks__", "**/*.js"],
+  "exclude": ["dist", "node_modules", "**/__mocks__", "**/*.js", "src/lib/asset-service/service/generatedAssetData.json"],
```
Comment (gomesalexandre) [inline tsconfig.json:38]: Confirmed app still builds and functions fine, and deploys to cloudflare fine on gome.
On the flip side, tsc should be blazingly faster than in develop now vs. borked with this PR.

Expect a follow-up to remove base64 shenanigans - we have none of them in develop as we speak, and this is absolutely not working , see added assets with identicons here

image
image

**Review by NeOMakinG — APPROVED** [review]
image

Also found assets for Base, Optimism, Polygon, Gnosis, Arb... Comparing with https://app.portals.fi/ looks sane so far

Ex:
image

I noticed the same thing than Apotheosis with the global search btw

**Pull request merged**